File: network/src/main/kotlin/org/bitlap/network/core/operation/Operation.kt
package org.bitlap.network.core.operation

import org.bitlap.common.logger
import org.bitlap.network.QueryResult
import org.bitlap.network.core.OperationType
import org.bitlap.network.core.RowSet
import org.bitlap.network.core.Session
import org.bitlap.network.core.TableSchema

/**
 * This class should abstract according to different operations, and each operation is a special implementation.
 *
 * @author 梦境迷离
 * @since 2021/9/5
 * @version 1.0
 */
abstract class Operation(val parentSession: Session, val opType: OperationType, val hasResultSet: Boolean = false) {
    protected val log = logger { }

    val opHandle: OperationHandle by lazy { OperationHandle(opType, hasResultSet) }

    lateinit var statement: String
    var confOverlay: Map<String, String>? = mutableMapOf()

    abstract fun run()

    fun remove(operationHandle: OperationHandle) {
        cache.remove(operationHandle)
    }

    protected val cache: MutableMap<OperationHandle, QueryResult> = mutableMapOf()

    fun getNextResultSet(): RowSet {
        return cache[opHandle]?.rows ?: RowSet()
    }

    fun getResultSetSchema(): TableSchema {
        return cache[opHandle]?.tableSchema ?: TableSchema()
    }
}

File: bitlap-server/src/main/scala/org/bitlap/server/rpc/Operation.scala
/* Copyright (c) 2022 bitlap.org */
package org.bitlap.server.rpc

import com.typesafe.scalalogging.LazyLogging
import org.bitlap.network.handles.OperationHandle
import org.bitlap.network.models.RowSet
import org.bitlap.network.{ models, OperationType }

import scala.collection.mutable

/** @author
 *    梦境迷离
 *  @since 2021/11/20
 *  @version 1.0
 */
abstract class Operation(
  val parentSession: Session,
  val opType: OperationType,
  val hasResultSet: Boolean = false
) extends LazyLogging {

  private var statement: String = _

  // super不能用于字段
  def getStatement: String = statement

  def setStatement(statement: String): Unit =
    this.statement = statement

  def getOpHandle: OperationHandle = opHandle

  lazy val opHandle: OperationHandle =
    new OperationHandle(opType, hasResultSet)
  lazy val confOverlay: mutable.Map[String, String] =
    mutable.HashMap[String, String]()

  protected lazy val cache: mutable.HashMap[OperationHandle, models.QueryResult] =
    mutable.HashMap()

  def run()

  def remove(operationHandle: OperationHandle) {
    cache.remove(operationHandle)
  }

  def getNextResultSet(): RowSet =
    cache.get(opHandle).map(_.rows).getOrElse(models.RowSet())

  def getResultSetSchema(): models.TableSchema =
    cache.get(opHandle).map(_.tableSchema).getOrElse(models.TableSchema())
}
